datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -270,37 +270,51 @@ fn not_in_list_utf8<OffsetSize: OffsetSizeTrait>(
 }
 
 //check all filter values of In clause are static.
-//include `CastExpr + Literal` or `Literal`
+//include `TryCastExpr/CastExpr + Literal` or `Literal`
 fn check_all_static_filter_expr(list: &[Arc<dyn PhysicalExpr>]) -> bool {
+    // TODO optimize the checker for constant in datafusion
```

Comment:
for datafusion, we need to add interface which is used to identify the expr is constant or  literal or not.
If we have this interface, we can make more optimization.
But the interface should be better in the logical expr not in the physical expr like sprak `fold`

Reply:
I think `ColumnarValue` is one such interface

https://github.com/apache/arrow-datafusion/blob/41b4e491663029f653e491b110d0b5e74d08a0b6/datafusion/expr/src/columnar_value.rs#L29-L34

So if you call `evaluate()` on a `Literal` `PhysicalExpr` the returned value is `ColumnarValue::Scalar`:

https://github.com/apache/arrow-datafusion/blob/master/datafusion/physical-expr/src/expressions/literal.rs#L72

Reply:
One way to make this code more general might be to change the check for all static expressions to *after* `evaluate()` is called. 

Specifically, maybe the  

```rust
        if list.len() > OPTIMIZER_INSET_THRESHOLD && check_all_static_filter_expr(&list) {
```

check could be refactored to call `evaluate` on all the elements of the list and then if all elements were `ColumnarValue::Scalar` and the list was sufficiently larger the set optimization could be invoked 🤔